File: ollama/codigo-final/mandelbrot/stable-code-3b/cot/temp_0/mandelbrot.py
import sys

def mandelbrot(width, height):
    # Initialize the image with all white pixels
    image = [[255] * width for _ in range(height)]

    for x in range(width):
        for y in range(height):
            # Calculate the complex number C based on the coordinate mapping
            real_C = 2.0 * x / (width - 1) - 1.5
            imag_C = 2.0 * y / (height - 1) - 1.0

            # Initialize Z to the initial value
            Z = (0, 0)

            # Perform the Mandelbrot set calculation for each point
            for _ in range(50):
                # Calculate Z' = Z\*Z + C
                Z_squared = Z[0] ** 2 - Z[1] ** 2
                Z_prime = (Z_squared * real_C - imag_C * Z[0] * Z[1]),
                               (Z_squared * imag_C + real_C * Z[0] * Z[1]))

                # Check if |Z'|² <= lim after each iteration
                if abs(Z_prime[0]) ** 2 + abs(Z_prime[1]) ** 2 <= lim:
                    break

                # Update Z to Z'
                Z = Z_prime

            # Set the pixel color based on whether the point belongs to the set or not
            if abs(Z[0]) ** 2 + abs(Z[1]) ** 2 <= lim:
                image[y][x] = [0, 0, 255]  # Blue pixel for in set
            else:
                image[y][x] = [255, 255, 255]  # White pixel for not in set

    return image

def write_pbm(width, height):
    header = f"P4\n{width} {height}\n"
    pixel_data = "".join([f"{i:08b}" for i in [0] * (width // 8) + [1] * width])

    output = header + pixel_data
    print(output)

if __name__ == "__main__":
    N = int(sys.argv[1]))
    image = mandelbrot(N, N))
    write_pbm(N, N))

File: ollama/codigo-final/mandelbrot/stable-code-3b/cot/temp_0/mandelbrot.cpp
#include <iostream>
#include <cmath>

const int N = 200; // Command-line argument parsing for N
const int MAX_ITERATIONS = 50; // Fixed iteration limit (50)
const double ESCAPE_THRESHOLD = 2.0 * 2.0; // Fixed escape threshold (lim = 2.0 * 2.0)

void mandelbrot(int width, int height) {
    std::cout << "P4\n";
    std::cout << N << " " << N << "\n";

    double x_min = -1.5;
    double x_max = 0.5;
    double y_min = -1.0;
    double y_max = 1.0;

    for (int y = 0; y < height; ++y) {
        double py = y * (y_max - y_min) / (height - 1) + y_min;

        for (int x = 0; x < width; ++x) {
            double px = x * (x_max - x_min) / (width - 1) + x_min;

            double Zr = 0.0; // Initialize real part of complex number Z
            double Zi = 0.0; // Initialize imaginary part of complex number Z

            const double C_REAL = 2.0 * px / N - 1.5; // Calculate real part of complex number C
            const double C_IMAGINARY = 2.0 * py / N - 1.0; // Calculate imaginary part of complex number C

            double ZrSquared = 0.0; // Initialize squared magnitude of complex number Z
            double ZiSquared = 0.0; // Initialize squared magnitude of complex number Z

            Zr = 0.0; // Reset real part of complex number Z
            Zi = 0.0; // Reset imaginary part of complex number Z

            for (int i = 0; i < MAX_ITERATIONS; ++i) {
                double new_Zr = Zr * Zr - C_REAL + C_IMAGINARY * Zi; // Calculate real part of complex number Z after one iteration
                double new_Zi = 2.0 * Zr * Zi + C_IMAGINARY * (1.0 - C_REAL / (N - 1)) - C_IMAGINARY * (C_REAL / (N - 1)) - C_IMAGINARY * (1.0 - C_REAL / (N - 1))) / (ZrSquared + ZiSquared); // Calculate imaginary part of complex number Z after one iteration

                if ((new_Zr * new_Zr) + (new_Zi * new_Zi) > ESCAPE_THRESHOLD) {
                    std::cout << "1 "; // Set pixel to white (1)
                    break;
                }

                Zr = new_Zr; // Update real part of complex number Z
                Zi = new_Zi; // Update imaginary part of complex number Z
            }

            if ((new_Zr * new_Zr) + (new_Zi * new_Zi) <= ESCAPE_THRESHOLD) {
                std::cout << "0 "; // Set pixel to black (0)
            }
        }
    }
    std::cout << "\n";
}

int main() {
    mandelbrot(N, N));
    return 0;
}
